src/main/java/com/digitalasset/testing/ledger/SandboxCommunicator.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright 2019 Digital Asset (Switzerland) GmbH and/or its affiliates
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.digitalasset.testing.ledger;
+
+import com.daml.ledger.rxjava.DamlLedgerClient;
+import com.digitalasset.ledger.api.v1.LedgerIdentityServiceGrpc;
+import com.digitalasset.ledger.api.v1.LedgerIdentityServiceOuterClass;
+import com.digitalasset.ledger.api.v1.testing.TimeServiceGrpc;
+import com.digitalasset.testing.ledger.clock.SandboxTimeProvider;
+import com.digitalasset.testing.store.DefaultValueStore;
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.time.Duration;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Consumer;
+
+import static com.digitalasset.testing.utils.SandboxUtils.getSandboxPort;
+import static com.digitalasset.testing.utils.SandboxUtils.waitForSandbox;
+
+public class SandboxCommunicator {
+  private static final Logger logger = LoggerFactory.getLogger(SandboxCommunicator.class);
+  private static final String COMPILATION_LOG = "integration-test-compilation.log";
+  private static final String DAML_EXE = "daml";
+  private int sandboxPort;
+
+  private final Optional<String> testModule;
+  private final Optional<String> testScenario;
+  private final Duration waitTimeout;
+  private final String[] parties;
+  private final Path darPath;
+  private final Consumer<DamlLedgerClient> setupApplication;
+
+  private SandboxRunner sandboxRunner;
+  private DamlLedgerClient ledgerClient;
+  private DefaultLedgerAdapter ledgerAdapter;
+  private ManagedChannel channel;
+
+  public SandboxCommunicator(
+      Optional<String> testModule,
+      Optional<String> testScenario,
+      Duration waitTimeout,
+      String[] parties,
+      Path darPath,
+      Consumer<DamlLedgerClient> setupApplication) {
+    this.testModule = testModule;
+    this.testScenario = testScenario;
+    this.waitTimeout = waitTimeout;
+    this.parties = parties;
+    this.darPath = darPath;
+    this.setupApplication = setupApplication;
+  }
+
+  public DamlLedgerClient getClient() {
+    return ledgerClient;
+  }
+
+  public DefaultLedgerAdapter getLedgerAdapter() {
+    return ledgerAdapter;
+  }
+
+  public ManagedChannel getChannel() {
+    return channel;
+  }
+
+  public void startSandbox() throws IOException, TimeoutException {
```

Comment:
These should be private. The above suggested public methods should call these.

Reply:
Done.